File: 2016/15.cs
namespace _2016;

public abstract class _15
{
    private static readonly string[] Data;

    static _15() => Data = Task.Run(() => Utils.GetInputData(15)).Result
        .Split('\n', StringSplitOptions.RemoveEmptyEntries);

    private class Disc(int positions, int initialPosition)
    {
        public int Positions { get; } = positions;
        public int InitialPosition { get; } = initialPosition;
    }

    private static List<Disc> ParseDiscs() => Data.Select(line =>
    {
        var parts = line.Split([' ', '#', ';', '.'], StringSplitOptions.RemoveEmptyEntries);
        var positions = int.Parse(parts[3]);
        var initialPosition = int.Parse(parts.Last());
        return new Disc(positions, initialPosition);
    }).ToList();

    private static bool IsSuccessfulDrop(List<Disc> discs, int time)
    {
        for (var i = discs.Count - 1; i >= 0; i--)
        {
            var disc = discs[i];
            var discPositionAtTime = (disc.InitialPosition + time + i + 1) % disc.Positions;
            if (discPositionAtTime != 0) return false;
        }
        return true;
    }

    private static int FindFirstSuccessfulDropTime(IEnumerable<Disc> discs)
    {
        var time = 0;
        var discsList = discs.ToList();

        while (true)
        {
            if (IsSuccessfulDrop(discsList, time)) return time;
            time++;
        }
    }

    private static int SolvePartOne() => FindFirstSuccessfulDropTime(ParseDiscs());

    private static int SolvePartTwo()
    {
        var discs = ParseDiscs();
        discs.Add(new Disc(11, 0));
        return FindFirstSuccessfulDropTime(discs);
    }

    public static void Run()
    {
        Console.WriteLine($"Part One: {SolvePartOne()}");
        Console.WriteLine($"Part Two: {SolvePartTwo()}");
    }
}

File: 2017/15.fs
namespace _2017

open System

module _15 =
    let Data = (Utils.GetInputData 15).Split('\n', StringSplitOptions.RemoveEmptyEntries)

    let parseData(prefix: string) = Data |> Array.find (_.StartsWith(prefix)) |> fun line -> line.Split(' ').[4] |> int64

    let generateNextValue (previousValue, factor) = (previousValue * factor) % 2147483647L

    let lowest16Bits value = value &&& 0xFFFFL

    let countMatchingPairs iterations filterA filterB =
        let mutable count = 0
        let mutable valueA = parseData "Generator A starts with"
        let mutable valueB = parseData "Generator B starts with"
        let mutable validPairs = 0
        while validPairs < iterations do
            valueA <- generateNextValue (valueA, 16807L)
            if filterA valueA then
                valueB <- generateNextValue (valueB, 48271L)
                while not (filterB valueB) do valueB <- generateNextValue (valueB, 48271L)
                if (lowest16Bits valueA) = (lowest16Bits valueB) then count <- count + 1
                validPairs <- validPairs + 1
        count

    let solvePartOne() = countMatchingPairs 40000000 (fun _ -> true) (fun _ -> true)

    let solvePartTwo() = countMatchingPairs 5000000 (fun value -> value % 4L = 0L) (fun value -> value % 8L = 0L)

    let Run() =
        printfn $"Part One: {solvePartOne()}"
        printfn $"Part Two: {solvePartTwo()}"